### PR TITLE
Add Google Spanner Support (google_spanner_instance)

### PR DESCRIPTION
**Pull request by nickithewatt**
This PR aims to addresses https://github.com/terraform-providers/terraform-provider-google/issues/67 which looks to add Google Spanner Support, specifically adding a new resource `google_spanner_instance`. CC @rileykarson @danawillow 

In order to make these PR's easier to digest and review, instead of creating one big PR with all Spanner support in it, I have created one for each new resource to be added. This is the first which adds support for the foundational `google_spanner_instance` resource. A separate one (https://github.com/terraform-providers/terraform-provider-google/pull/271) adds support for creating the Database within the Instance, i.e. the `google_spanner_database` resource.

### CODE REVIEW

File: google/resource_spanner_instance.go [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+			"num_nodes": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  1,
+			},
+
+			"labels": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				ForceNew: true,
```
Comment (rosbo) [inline google/resource_spanner_instance.go]: The patch API support labels updates. Remove this `ForceNew` and add support for it in the update method below.

File: google/resource_spanner_instance.go [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+	fieldMask := []string{}
+	if d.HasChange("num_nodes") {
+		fieldMask = append(fieldMask, "nodeCount")
+		uir.Instance.NodeCount = int64(d.Get("num_nodes").(int))
+	}
+	if d.HasChange("display_name") {
+		fieldMask = append(fieldMask, "displayName")
+		uir.Instance.DisplayName = d.Get("display_name").(string)
+	}
+
```
Comment (rosbo) [inline google/resource_spanner_instance.go]: Add a block for changes to `labels`

File: google/resource_spanner_instance.go [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return &spannerInstanceId{
+		Project:  project,
+		Instance: d.Get("name").(string),
+	}, nil
+}
+
+func extractInstanceConfigFromApi(nameUri string) string {
```
Comment (rosbo) [inline google/resource_spanner_instance.go]: Consider renaming to `extractInstanceConfigFromUri(configUri string) string`

File: google/resource_spanner_instance.go [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+	return extractLastResourceFromUri(nameUri)
+}
+
+func extractInstanceNameFromApi(nameUri string) string {
```
Comment (rosbo) [inline google/resource_spanner_instance.go]: Consider renaming to `extractInstanceNameFromUri(nameUri string) string`

File: google/resource_spanner_instance_test.go [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+	"github.com/hashicorp/terraform/terraform"
+
+	"google.golang.org/api/googleapi"
+	"google.golang.org/api/spanner/v1"
+	"strings"
+)
+
+// Unit Tests
+
+func TestExtractInstanceConfigFromApi_withFullPath(t *testing.T) {
```
Comment (rosbo) [inline google/resource_spanner_instance_test.go]: Renaming test accordingly if you change the method name.

File: website/docs/r/spanner_instance.html.markdown [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+* `num_nodes` - (Optional, Computed) The number of nodes allocated to this instance.
+   Defaults to `1`. This can be updated after creation.
+
+* `project` - (Optional) The project in which the resource belongs. If it
+    is not provided, the provider project is used.
+
+## Attributes Reference
+
+No additional attributes are computed other than those defined above.
```
Comment (rosbo) [inline website/docs/r/spanner_instance.html.markdown]: Don't forget to add the `state` computed attribute in this section.

**Comment by nickithewatt** [comment]
Thanks, aiming to try and get round to these changes some time in the next few days.

File: google/resource_spanner_instance.go [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+				Optional: true,
+				ForceNew: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+
+			"project": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
```
Comment (nickithewatt) [inline google/resource_spanner_instance.go]: The project field is not actually computed (in line with most other resources), did you want this explicitly set it to the default false value?

**Comment by nickithewatt** [comment]
@rosbo thanks for review: replied to one comment above and other changes made as requested. Let me know if there is anything else that needs doing on this :)

**Comment by rosbo** [comment]
make testacc TESTARGS="-run TestAccGoogleSpannerInstance_" TEST="./google"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./google -v -run TestAccGoogleSpannerInstance_ -timeout 120m
=== RUN   TestAccGoogleSpannerInstance_importInstance
--- PASS: TestAccGoogleSpannerInstance_importInstance (76.53s)
=== RUN   TestAccGoogleSpannerInstance_importProjectInstance
--- PASS: TestAccGoogleSpannerInstance_importProjectInstance (25.50s)
PASS
ok  	github.com/terraform-providers/terraform-provider-google/google	102.157s

make testacc TESTARGS="-run TestAccSpannerInstance_" TEST="./google"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./google -v -run TestAccSpannerInstance_ -timeout 120m
=== RUN   TestAccSpannerInstance_basic
--- PASS: TestAccSpannerInstance_basic (12.77s)
=== RUN   TestAccSpannerInstance_basicWithAutogenName
--- PASS: TestAccSpannerInstance_basicWithAutogenName (11.35s)
=== RUN   TestAccSpannerInstance_duplicateNameError
--- PASS: TestAccSpannerInstance_duplicateNameError (11.62s)
=== RUN   TestAccSpannerInstance_update
--- PASS: TestAccSpannerInstance_update (21.88s)
PASS
ok  	github.com/terraform-providers/terraform-provider-google/google	57.754s

**Comment by rosbo** [comment]
Thank you so much for adding Spanner support to Terraform!

**Pull request merged**